### ISSUE TITLE
Bug: [no-useless-default-assignment] False positive when defaulting parameter to undefined

### ISSUE THREAD

**Issue by doberkofler** [typescript-eslint/typescript-eslint#11847]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=5.9.2&fileType=.tsx&code=GYVwdgxgLglg9mABAWwJ4DFzXmAFABwEYAuRAZygCcYwBzAGkXwCZSwRkAjAU0sQB9E4ACbdgNbsMQBeIWFHiwkgJSIA3gFgAUIkQQEZOABtuAOiNxaBQoxbKA3NoC%2B27WkyRYCXAHJCPh20gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYkACAXAngA4oDG0AlobgLQrzkB2uA9AwPbWzKJLLLUATRADMAhghqi%2B5AOYMAtoibooiaNDbRIAXxDagA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function myFunction(p1: string, p2: number | undefined = undefined) {
  console.log(p1, p2);
}

myFunction('1');
```

### ESLint Config

```javascript
module.exports = {
  "rules": {"@typescript-eslint/no-useless-default-assignment": "error"}
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

This use should get flagged.

### Actual Result

When removing the default, typescript will reports the error `2554: Expected 2 arguments, but got 1. 5:1 - 5:11`

**Comment by jimmy-zhening-luo** [comment]
It also false-positives on this default assignment of `b`, claiming that `b` is not optional even though TypeScript knows that `b` is possibly undefined and in fact refuses to compile if I try to assign it to a non-nullable:

```typescript
export function foo(
  bool1: boolean,
  bool2: boolean,
): string {
  const {
    a,
    b = "default", // Expected: OK; Actual: Default value is useless because the property is not optional. eslint(@typescript-eslint/no-useless-default-assignment)
  } = bool1
    ? bool2
      ? {
          a: Math.random(),
          b: "Hello",
        }
      : {
          a: Math.random(),
          b: "World",
        }
    : {
        a: Math.random(),
      };

  console.log(a);

  return b;
}
```

**Comment by ulrichstark** [comment]
Hey @jimmy-zhening-luo, thanks for reporting! Because your case is pretty different, I condensed it down and created a separate issue #11850.

**Comment by kirkwaiblinger** [comment]
Hmm, I guess `= undefined` is equivalent to making the property optional for TS type purposes...

([playground](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=GYVwdgxgLglg9mABAWwJ4DFzXmAFABwEYAuRAZygCcYwBzAGkXwCZSwRkAjAU0sQB9E4ACbdgNbsICUiAN4BYAFCJEEBGTgAbbgDpNcWgUKMWUgNxKAvkqVpMkWAlwByQs-O2MWR3mfNnjCJiEtKe9thOzgDMAYjMACweijaKoA44KF7pCAAiYgCGIJpQAAr5lPnIRqQU1HQmrIjsXLwCQmCi4mCSiAC87Z0hMgrKqupauvqGRA1J1smKdt44ecCFxWUVVa7uFotZEWCr66XllS7%2BgR3B3aH74T7HRadbLjGMCUkpaYeZDzgAeXwPnymk25yINSoNAYTGYAH42BweJRhkoVGowBptHoDEZZnt5mFlgggSCwWdtm4kktsmAyThQeDtpcBjdJFJiXSGQgmZS3rFPnslEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tmUSWWS0AJogBmAQwR1xAygHMmAW0Qt0UAO7joTSODABfEPqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false))

```ts
function myFunction(p1: string, p2?: number) {
  console.log(p1, p2);
}

myFunction('1');
myFunction('2', undefined)
myFunction('3', 24);
```

It's a little awkward not to flag because `= undefined` is basically useless from a JS perspective (_technically_ it affects the function's `.length`), and an optional parameter is equivalent for TS type purposes and saves 20 characters, so I would expect _a_ lint rule to complaint about this. 

I guess we have choices:

```
Is optional-parameters-vs-default-undefined-parameters in scope for this rule to flag?

if not in scope:
   Resolution is to not flag default-undefined parameters.

else:
   
   Is optional-parameters-vs-default-undefined-parameters a neutral, purely stylistic preference,
   or should we assert a preference for one or the other?

   if we assert a preference for optional parameters:
      resolution is to continue unconditionally flagging default-undefined parameters, 
      but add documentation and (suggestion?) fixer to convert to optional parameter
   else:
      resolution is that whether to flag optional parameters or default-undefined parameters
      should be controlled by an option, and corresponding documentation/fixer should be added
```

I think I'd personally most prefer the behavior of "yes in scope, and we assert a preference for optional parameters"... But I can see how that's debatable.

@typescript-eslint/triage-team

**Comment by JoshuaKGoldberg** [comment]
> so I would expect _a_ lint rule to complaint about this.

Yeah, me too.

> stylistic

I also think it's in scope for this rule - agreed it's more than just a purely stylistic preference. Making the parameter `?` optional when it has an `| undefined` type _is_ different at runtime than `= undefined`, as you noted.

I do think there should probably be a different report message for this case. Given that the `= undefined` is what's making the parameter optional, I think the messaging should be changed to say that the optionality is being added in an unnecessarily runtime-changing way, not that it's already optional.

> (suggestion?) fixer

+1 to a fixer. I think this should be safe as a full fix, not a suggestion. `= undefined` is just an extra verbose way to assign the parameter a value of `undefined`, right?

**Comment by kirkwaiblinger** [comment]
> I also think it's in scope for this rule - agreed it's more than just a purely stylistic preference. Making the parameter `?` optional when it has an `| undefined` type _is_ different at runtime than `= undefined`, as you noted.

To be clear the runtime difference isn't the reason that I'm arguing to prefer the optional parameter style. I just think the optional parameter style is much better because it's many less characters, and `= undefined` feels silly since it doesn't change the runtime value of the parameter. I only mention the technical caveat about the `.length` being affected to acknowledge its existence and then argue that we're safe to ignore it (because apart from [one, questionably designed API](https://expressjs.com/en/guide/error-handling.html#writing-error-handlers), `.length` is never used in practice and already clashes with the mindset of other eslint rules).

**Comment by reduckted** [comment]
Having a fixer change a default argument to an optional argument could cause [`default-param-last`](https://typescript-eslint.io/rules/default-param-last/) to start reporting a problem.

Should the fixer take that into consideration (even though it doesn't know of that rule is actually enabled), or is that just a case of "you've got conflicting rules, so you'll have to fix the `default-param-last` rule yourself"?

I'm not suggesting that it should do anything about it, but it's just a side effect that I saw with the current fixer implementation and thought I'd raise it. 😄

**Comment by kirkwaiblinger** [comment]
@reduckted Do you have an example of this? It [seems to be fine](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=GYVwdgxgLglg9mABAWwJ4DFzXmAFABwEYAuRAZygCcYwBzRAXkQHJg45mAaRfAJgH5SYEMgBGAU0oBKRAG8AsAChEiCAjJwANuIB0muLQKFufKQG4lAXyVK0mSLAS5mhZuaVA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYkACAXAngA4oDG0AlobgLQrzkB2uA9AwPbWzKJLLLUATRADMAhghqi%2B5AOYMAtoibooiaNDbRIAGnBQ8RUhSq1k9JsyFiJ1QqOij51eFNwrIAd3sNIAXxC%2BQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) as far as I can tell?

**Comment by reduckted** [comment]
Ah! Oprional parameters after default parameters is OK. The current fixer simply removes the default value rather than changing it to be optional which is why I was seeing `default-param-last` appear. 

It's good to know that the new fixer won't have to worry about that. 👍
